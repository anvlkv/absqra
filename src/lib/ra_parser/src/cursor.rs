use ra_lexer::token::{Token};
use ra_lexer::cursor::Position;
use core::iter::Peekable;

pub(crate) struct Cursor<'token, I> 
    where I: Iterator<Item = Token<'token>> 
{
    pub position: Position,
    pub level: u16,
    tokens: Peekable<I>,
    consumed_len: usize
}

impl <'token, I> Cursor<'token, I> 
where I: Iterator<Item = Token<'token>> 
{
    pub fn new(tokens: I) -> Self
        
    {
        Self {
            position: Position(0, 0),
            level: 0,
            tokens: tokens.peekable(),
            consumed_len: 0
        }
    }

    pub fn first_ahead(&mut self) -> Option<&Token<'token>> {
        self.tokens.peek()
    }

    pub fn bump(&mut self) -> Option<Token<'token>> {
        let tok = self.tokens.next();
        match tok {
            Some(token) => {
                self.position = token.position.1;
                self.level = token.level;
                self.consumed_len += 1;
                Some(token)
            }
            None => None
        }
    }

    pub fn is_eof(& mut self) -> bool {
        self.first_ahead() == None
    }

    // pub fn do_while<F, D>(&mut self, mut test: F, mut cb: D)
    //     where F: FnMut (u16, Position, TokenKind) -> bool,
    //     D: FnMut (Token)
    // {
    //     loop {
    //         let next_token = self.first_ahead();
    //         match next_token {
    //             Some(t) => {
    //                 if test(t.level, t.position.0, t.kind) {
    //                     cb(self.bump().unwrap())
    //                 }
    //                 else {
    //                     break
    //                 }
    //             },
    //             None => break
    //         }
    //     }
    // }

    // pub fn read_within(&mut self, open: TokenKind, close: TokenKind) -> Vec<Token> {
    //     let mut read: Vec<Token> = Vec::new();

    //     loop {
    //         let next_token = self.first_ahead();

    //         match next_token {
    //             Some(tok) => {
    //                 match tok.kind {
    //                     k if k == open => {
    //                         let open_token = self.bump().unwrap();
    //                         let mut nested = self.read_within(open_token.kind, close.clone());
    //                         read.append(&mut nested);
    //                     },
    //                     k if k == close => {
    //                         self.bump();
    //                         break
    //                     },
    //                     _ => read.push(self.bump().unwrap())
    //                 }
    //             },
    //             None => break
    //         }

    //         // println!("{:?}", &read);
    //     }

    //     read
    // }
}