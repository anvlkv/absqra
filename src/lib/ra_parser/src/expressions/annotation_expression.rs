use serde::Serialize;
use super::{ParserError, ParseByToken, RaToken, TokenKind};

#[derive(Serialize, Clone, Debug)]
pub struct AnnotationExpression {

}

impl ParseByToken for AnnotationExpression {
    fn new(token: RaToken) -> std::result::Result<Self, Vec<ParserError>> { 
        todo!("implement new")
    }
    fn append_token(self, token: RaToken) -> std::result::Result<Self, Vec<ParserError>> { 
        todo!("implement append_token")
    }
    fn allowed_tokens(&self) -> Vec<TokenKind> { 
        todo!("implement allowed_tokens")
    }

    fn starts_with_tokens() -> Vec<TokenKind<'static>> { 
        [
            
        ].to_vec()
     }

}

