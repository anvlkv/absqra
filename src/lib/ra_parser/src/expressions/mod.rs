pub (crate) mod buffered;


#[macro_use]
mod repeat_kindly_macros;

pub mod output_expression_kind;
pub mod output_expression;
pub mod input_expression;
pub mod reference_expression;
pub mod annotation_expression;
pub mod context_expression;
pub mod content_expression;
pub mod grouping_expression;
pub mod logic_operation;
pub mod logic_expression;
pub mod operation_expression;
pub mod procedure_expression;
pub mod math_operation;
pub mod expression;



use super::errors::ParserError;
use super::parsed_by_token::ParsedByToken;
use ra_lexer::token::{RaToken, TokenKind};
use failure::Backtrace;
