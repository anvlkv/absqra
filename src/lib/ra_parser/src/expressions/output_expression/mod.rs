use super::*;

pub mod output_expression;
pub mod expression_member;


use expression_member::OutputExpressionMember;
pub use output_expression::OutputExpression;