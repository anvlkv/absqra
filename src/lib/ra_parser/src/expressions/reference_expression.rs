use super::*;

#[derive(Serialize, Clone, Debug)]
pub struct ReferenceExpression {

}

impl<'a> ParsedByToken<'a, ReferenceExpression> for ReferenceExpression {
    fn new(token: RaToken<'a>) -> Result<Box<ReferenceExpression>, Vec<ParserError>> { 
        todo!("implement new")
    }
    fn append_token(self, token: RaToken<'a>) -> Result<Box<ReferenceExpression>, Vec<ParserError>> { 
        todo!("implement append_token")
    }
    fn allowed_tokens(&self) -> Vec<TokenKind<'a>> { 
        todo!("implement allowed_tokens")
    }

    fn starts_with_tokens() -> Vec<TokenKind<'static>> { 
        vec![
            TokenKind::At,
            TokenKind::Identifier(Default::default())
        ]
     }

}

