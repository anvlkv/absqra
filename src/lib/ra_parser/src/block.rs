use crate::errors::ParserError;
use crate::expressions::expression::Expression;
use crate::parsed_by_token::{ParsedByToken, StartsWithTokens};
use failure::Backtrace;
use ra_lexer::cursor::Position;
use ra_lexer::token::{RaToken, TokenKind};
use serde::Serialize;

#[derive(Serialize, Clone, Debug)]
pub enum BlockKind<'a> {
    Program,
    Expression(Expression<'a>),
    Union(usize),
}

#[derive(Serialize, Clone, Debug)]
pub struct Block<'a> {
    pub position: (Position, Position),
    pub level: u16,
    pub kind: BlockKind<'a>,
    pub children: Vec<Box<Block<'a>>>,
}

impl<'a> Block<'a> {
    pub fn new_program() -> Result<Box<Self>, Vec<ParserError>> {
        Ok(Box::new(Self {
            kind: BlockKind::Program,
            level: 0,
            position: (Position::default(), Position::default()),
            children: Vec::new(),
        }))
    }

    pub fn new_union(level: u16) -> Result<Box<Self>, Vec<ParserError>> {
        Ok(Box::new(Self {
            kind: BlockKind::Union(0),
            level,
            position: (Position::default(), Position::default()),
            children: Vec::new(),
        }))
    }

    fn belongs_to_primary_expression(&self, token: &RaToken<'a>) -> bool {
        match &self.kind {
            BlockKind::Program => false,
            BlockKind::Union(_) => (self.position.0).0 == (token.position.0).0,
            BlockKind::Expression(exp) => (exp.position.0).0 == (token.position.0).0
        }
    }

    fn belongs_to_sibling_expression(&self, token: &RaToken<'a>) -> bool {
        match &self.kind {
            BlockKind::Program => false,
            BlockKind::Union(_) => {
                if self.children.last().is_some() {
                    self.children.last().unwrap().belongs_to_sibling_expression(token)
                }
                else {
                    false
                }
            },
            BlockKind::Expression(exp) => exp.level == token.level && !self.belongs_to_primary_expression(token)
        }
    }

    fn belongs_to_child_expression(&self, token: &RaToken<'a>) -> bool {
        match &self.kind {
            BlockKind::Program => true,
            BlockKind::Union(_) => {
                if self.children.last().is_some() {
                    self.children.last().unwrap().belongs_to_child_expression(token)
                }
                else {
                    false
                }
            },
            BlockKind::Expression(exp) => exp.level > token.level
        }
    }

    fn append_token_child(self, token: RaToken<'a>) -> Result<Box<Self>, Vec<ParserError>> {
        let mut children = self.children.clone();
        if children.len() > 0 {
            let child = children.last().unwrap().clone().append_token(token)?;
            children.pop();
            children.push(child);
            Ok(Box::new(Self {
                children,
                position: (self.position.0, token.position.1),
                ..self
            }))
        } else {
            children.push(Self::new_from_token(token)?);
            Ok(Box::new(Self {
                children,
                position: (self.position.0, token.position.1),
                ..self
            }))
        }
    }
}

impl<'a> ParsedByToken<'a> for Block<'a> {
    fn new_from_token(token: RaToken<'a>) -> Result<Box<Self>, Vec<ParserError>> {
        let expression = Expression::new_from_token(token)?;
        Ok(Box::new(Self {
            kind: BlockKind::Expression(*expression),
            level: token.level,
            position: token.position.clone(),
            children: Vec::new(),
        }))
    }

    fn append_token(self, token: RaToken<'a>) -> Result<Box<Self>, Vec<ParserError>> {
        if self.belongs_to_primary_expression(&token) {
            match self.kind {
                BlockKind::Expression(old_expression) => {
                    let new_expression = old_expression.append_token(token)?;
        
                    Ok(Box::new(Self{
                        kind: BlockKind::Expression(*new_expression),
                        position: (self.position.0, token.position.1),
                        ..self
                    }))
                },
                BlockKind::Union(_) => {
                    self.append_token_child(token)
                },
                BlockKind::Program => Err(vec![ParserError::InvalidBlock(Backtrace::new())])
            }
        }
        else if self.belongs_to_child_expression(&token) {
            self.append_token_child(token)
        }
        else {
            Err(vec![ParserError::InvalidBlock(Backtrace::new())])
        }
    }

    fn allowed_tokens(&self) -> Vec<TokenKind<'a>> {
        match &self.kind {
            BlockKind::Expression(expression) => expression.allowed_tokens(),
            _ => Expression::starts_with_tokens(),
        }
    }

    fn required_tokens(&self) -> Vec<TokenKind<'a>> {
        match &self.kind {
            BlockKind::Expression(expression) => expression.required_tokens(),
            _ => Expression::starts_with_tokens(),
        }
    }
}

impl<'a> StartsWithTokens<'a> for Block<'a> {
    fn starts_with_tokens() -> Vec<TokenKind<'static>> {
        Expression::starts_with_tokens()
    }
}
