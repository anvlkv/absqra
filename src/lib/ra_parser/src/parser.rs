use crate::errors::ParserError;
use crate::block::Block;
use crate::parsed_by_token::ParsedByToken;
use ra_lexer::tokenize;
use ra_lexer::token::TokenKind;

pub fn parse<'a>(input: &'a str) -> Result<Block, (Vec<ParserError>, Option<Block>)> {
    let mut errors: Vec<ParserError> = Vec::new();
    let mut tokens_stream = tokenize(input);
    
    let mut block = {
        match Block::new_program() {
            Ok(b) => b,
            Err(e) => {
                errors.extend(e);
                return Err((errors, None));
            }
        }
    };

    while let Some(token_result) = tokens_stream.next() {
        match token_result {
            Ok(token) => {
                if token.kind == TokenKind::Comment {
                    continue;
                }
                else {
                    match block.clone().append_token(token) {
                        Ok(blk) => block=blk,
                        Err(e) => errors.extend(e)
                    }
                }
            },
            Err(e) => {
                errors.push(e.into());
            }
        }
    }

    if errors.len() == 0 {
        Ok(*block)
    }
    else {
        Err((errors, Some(*block)))
    }
}