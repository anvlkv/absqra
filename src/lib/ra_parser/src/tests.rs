mod lib {
    use ra_lexer::cursor::Position;
    use ra_lexer::token::{Token, TokenKind};

    use crate::block::BlockKind;
    use crate::expressions::output_expression::ExpressionMember;
    use crate::expressions::output_expression::{MathOperation, OperationKind, OutputExpression};
    use crate::expressions::reference_expression::ReferenceExpression;
    use crate::parser::parse;

    #[test]
    fn it_should_return_program_block() {
        assert_eq!(parse("abc").expect("can't parse").kind, BlockKind::Program);
    }

    #[test]
    fn it_should_parse_block_level_expressions() {
        let program = parse("abc + 2").expect("can't parse");
        let parsed = program.children.iter().nth(0).unwrap();
        assert_eq!(
            parsed.kind,
            BlockKind::Output(OutputExpression(
                Box::new(ExpressionMember::ReferenceExpression(ReferenceExpression(Token {
                    kind: Some(TokenKind::Identifier("abc")),
                    position: (Position(1, 0), Position(1, 3)),
                    len: 3,
                    content: "abc",
                    level: 0,
                }, None))),
                Some(OperationKind::MathOperation(MathOperation::Sum)),
                Some(Box::new(ExpressionMember::Literal(Token {
                    kind: Some(TokenKind::Int(2)),
                    position: (Position(1, 6), Position(1, 7)),
                    level: 0,
                    content: "2",
                    len: 1
                })))
            ))
        )
        // assert_eq!(parsed.expression.1, Some(OperationKind::MathOperation(MathOperation::Sum)));

        // assert_eq!(
        //     parsed.expression.0,
        //     Token {
        //         content: "abc",
        //         kind: Some(TokenKind::Identifier("abc")),
        //         len: 3,
        //         level: 0,
        //         position: (Position(1, 0), Position(1, 3))
        //     }
        // );

        // assert_eq!(
        //     (parsed.expression.2.as_ref().unwrap())
        //     Token {
        //         content: "2",
        //         kind: Some(TokenKind::Int(2)),
        //         len: 1,
        //         level: 0,
        //         position: (Position(1, 6), Position(1, 7))
        //     }
        // );
    }

    use ra_dev_tools::insta::{assert_json_snapshot, assert_snapshot};
    
    use std::fs::File;
    use std::io::Read;
    use ra_dev_tools::for_each_ra_example_file;

    #[test]
    fn it_should_match_snapshots() {
        println!("{:?}", std::env::args());
        for_each_ra_example_file(|example| {
            let mut file = File::open(example.path()).unwrap();
            
            let mut contents = String::new();
            file.read_to_string(&mut contents).unwrap();

            let block_tree = {
                match parse(&contents) {
                    Ok(b) => b,
                    Err((errors, parsed)) => {

                        let formatted_errors: String = {
                            errors.into_iter().map(|e| format!("{}", e)).collect::<Vec<String>>().join("\n")
                        };

                        
                        assert_snapshot!(String::from(example.path().to_str().unwrap()) + "__ERR__MESSAGES__", formatted_errors);
                        assert_json_snapshot!(String::from(example.path().to_str().unwrap()) + "__ERR__", parsed);


                        panic!("failed to parse example {:?}", example.path());
                    }
                }
            };
            assert_json_snapshot!(example.path().to_str().unwrap(), block_tree)
        })       
    }
}
