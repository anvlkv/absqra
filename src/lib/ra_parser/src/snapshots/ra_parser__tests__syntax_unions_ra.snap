---
source: src/lib/ra_parser/src/tests.rs
expression: ast
---
[
  [
    "Root"
  ]
]
